src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,275 @@
+          curGroupKey = new Tuple(groupSchema);
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            gTypes[gKey] = tb.getSchema().getColumnType(gFields[gKey]);
+            switch (gTypes[gKey]) {
+              case BOOLEAN_TYPE:
+                curGroupKey.set(gKey, tb.getBoolean(gFields[gKey], row));
+                break;
+              case STRING_TYPE:
+                curGroupKey.set(gKey, tb.getString(gFields[gKey], row));
+                break;
+              case DATETIME_TYPE:
+                curGroupKey.set(gKey, tb.getDateTime(gFields[gKey], row));
+                break;
+              case INT_TYPE:
+                curGroupKey.set(gKey, tb.getInt(gFields[gKey], row));
+                break;
+              case LONG_TYPE:
+                curGroupKey.set(gKey, tb.getLong(gFields[gKey], row));
+                break;
+              case FLOAT_TYPE:
+                curGroupKey.set(gKey, tb.getFloat(gFields[gKey], row));
+                break;
+              case DOUBLE_TYPE:
+                curGroupKey.set(gKey, tb.getDouble(gFields[gKey], row));
+                break;
+            }
+          }
+        } else if (!TupleUtils.tupleEquals(tb, gFields, row, curGroupKey, gRange, 0)) {
+          /* Different grouping key than current one, flush current agg result to result buffer. */
+          addToResult();
+          // store current group key as a tuple
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            switch (gTypes[gKey]) {
+              case BOOLEAN_TYPE:
+                curGroupKey.set(gKey, tb.getBoolean(gFields[gKey], row));
+                break;
+              case STRING_TYPE:
+                curGroupKey.set(gKey, tb.getString(gFields[gKey], row));
+                break;
+              case DATETIME_TYPE:
+                curGroupKey.set(gKey, tb.getDateTime(gFields[gKey], row));
+                break;
+              case INT_TYPE:
+                curGroupKey.set(gKey, tb.getInt(gFields[gKey], row));
+                break;
+              case LONG_TYPE:
+                curGroupKey.set(gKey, tb.getLong(gFields[gKey], row));
+                break;
+              case FLOAT_TYPE:
+                curGroupKey.set(gKey, tb.getFloat(gFields[gKey], row));
+                break;
+              case DOUBLE_TYPE:
+                curGroupKey.set(gKey, tb.getDouble(gFields[gKey], row));
+                break;
+            }
+          }
+          reinitializeAggStates();
+        }
+        // update aggregator states with current tuple
+        for (int agg = 0; agg < aggregators.length; ++agg) {
+          aggregators[agg].addRow(tb, row, aggregatorStates[agg]);
+        }
+      }
+      tb = child.nextReady();
+    }
+
+    /*
+     * We know that child.nextReady() has returned <code>null</code>, so we have processed all tuple we can. Child is
+     * either EOS or we have to wait for more data.
+     */
+    if (child.eos()) {
```

Comment:
Why can't you return whatever aggregation results you currently have available? Since you don't persist aggregation results to your `TupleBuffer` until you know the current group key will never be seen again, why can't you finalize the current `TupleBatch` and return that? I think that's what you have to do for this to be called "streaming"...